src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -2288,168 +2288,239 @@ public IndexDefinition GetIndexDefinition()
+                                resultToFill.AddExplanation(document.Explanation);
+
+                            using (gatherScope?.Start())
+                                includeDocumentsCommand.Gather(document.Result);
+
+                            includeCountersCommand?.Fill(document.Result);
                         }
+                    }
+                    catch (Exception e)
+                    {
+                        if (resultToFill.SupportsExceptionHandling == false)
+                            throw;
 
-                        FillQueryResult(resultToFill, isStale, query.Metadata, documentsContext, indexContext);
+                        resultToFill.HandleException(e);
+                    }
 
-                        using (var reader = IndexPersistence.OpenIndexReader(indexTx.InnerTransaction))
-                        {
-                            using (var queryScope = query.Timings?.For(nameof(QueryTimingsScope.Names.Query)))
-                            {
-                                QueryTimingsScope gatherScope = null;
-                                QueryTimingsScope fillScope = null;
+                    using (fillScope?.Start())
+                        includeDocumentsCommand.Fill(resultToFill.Includes);
 
-                                if (queryScope != null && query.Metadata.Includes?.Length > 0)
-                                {
-                                    var includesScope = queryScope.For(nameof(QueryTimingsScope.Names.Includes), start: false);
-                                    gatherScope = includesScope.For(nameof(QueryTimingsScope.Names.Gather), start: false);
-                                    fillScope = includesScope.For(nameof(QueryTimingsScope.Names.Fill), start: false);
-                                }
+                    if (includeCountersCommand != null)
+                        resultToFill.AddCounterIncludes(includeCountersCommand);
+
+                    resultToFill.TotalResults = Math.Max(totalResults.Value, resultToFill.Results.Count);
+                    resultToFill.SkippedResults = skippedResults.Value;
+                    resultToFill.IncludedPaths = query.Metadata.Includes;
+                }
+
+                return;
+            }
+        }
 
-                                var totalResults = new Reference<int>();
-                                var skippedResults = new Reference<int>();
-                                IncludeCountersCommand includeCountersCommand = null;
+        protected class QueryContext : IDisposable
+        {
+            public TransactionOperationContext Context;
+            public bool IsStale;
+            public IDisposable ReturnContextToPool { get; set; }
 
-                                var fieldsToFetch = new FieldsToFetch(query, Definition);
+            public void Dispose()
+            {
+                using (ReturnContextToPool)
+                {
+                    Context.CloseTransaction();
```

Comment:
IIRC, doesn't this already happen?

Reply:
As it was implemented so far the read transaction is closed after we retrieve all the ids
I just refactoring this so I can reuse the same code in query for documents and in query for ids